### PR TITLE
Fix mypy typing errors in optimizer loop

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #7037
Fixes mypy "typing" on the new optimizer loop because what else would I do on the weekend. 

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

NO.

### CODE REVIEW

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 1]
```diff
@@ -236,13 +251,15 @@ def _run_optimization_start(self, opt_idx: int, optimizer: torch.optim.Optimizer
             optimizer: the optimizer to use
 
         """
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:254]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 2]
```diff
@@ -62,21 +62,22 @@ def connect(self, **kwargs: "Loop") -> None:
         raise NotImplementedError(f"{self.__class__.__name__} does not connect any child loops.")
 
     def reset(self) -> None:
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:65]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 3]
```diff
@@ -94,18 +95,19 @@ def on_run_end(self) -> Tuple[_OUTPUTS_TYPE, Optional[Any]]:
...
         """Performs the backward step.
 
         Args:
             loss: The loss value to back-propagate on
-            optimizer: Current optimizer being used. ``None`` if using manual optimization.
-            opt_idx: Index of the current optimizer being used. ``None`` if using manual optimization.
+            optimizer: Current optimizer being used
+            opt_idx: Index of the current optimizer being used
         """
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:110]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 4]
```diff
@@ -118,19 +120,20 @@ def backward(
...
         Args:
-            batch_idx: the index of the current batch
             split_batch: the current tbptt split of the whole batch
-            opt_idx: the index of the current optimizer or `None` in case of manual optimization
-            optimizer: the current optimizer or `None` in case of manual optimization
+            batch_idx: the index of the current batch
+            optimizer: the current optimizer
+            opt_idx: the index of the current optimizer
         """
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:136]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 5]
```diff
@@ -177,6 +184,7 @@ def _make_closure(
         Build a closure object that captures the given arguments and runs the `training_step` function and optionally
         other functions such as `backward` and `zero_grad`.
         """
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:187]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 6]
```diff
@@ -217,6 +229,7 @@ def _make_backward_fn(
         """
 
         def backward_fn(loss: Tensor):
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:232]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 7]
```diff
@@ -236,13 +251,15 @@ def _run_optimization_start(self, opt_idx: int, optimizer: torch.optim.Optimizer
...
         """
+
         # make sure only the gradients of the current optimizer's parameters are calculated
         # in the training step to prevent dangling gradients in multiple-optimizer setup.
         if len(self.trainer.optimizers) > 1:
             model = self.trainer.lightning_module
             model.toggle_optimizer(optimizer, opt_idx)
 
     def _run_optimization_end(self, opt_idx: int) -> None:
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:262]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 8]
```diff
@@ -259,10 +276,11 @@ def _optimizer_step(
             train_step_and_backward_closure: the closure function performing the train step and computing the
                 gradients. By default called by the optimizer (if possible)
         """
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:279]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 9]
```diff
@@ -296,6 +314,7 @@ def _on_before_zero_grad(self, optimizer: torch.optim.Optimizer) -> None:
         Args:
             optimizer: the current optimizer
         """
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:317]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 10]
```diff
@@ -308,6 +327,7 @@ def _optimizer_zero_grad(self, batch_idx: int, optimizer: torch.optim.Optimizer,
             optimizer: the current optimizer
             opt_idx: the index of the current optimizer
         """
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:330]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 11]
```diff
@@ -325,6 +345,7 @@ def _training_step(
         Returns:
             an AttributeDict containing the loss value and the training step output.
         """
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:348]:
```suggestion
```

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 12]
```diff
@@ -362,6 +383,7 @@ def _track_and_norm_grad(self, optimizer: torch.optim.Optimizer) -> Dict[str, fl
         Args:
             optimizer: the current optimizer
         """
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:386]:
```suggestion
```

File: pytorch_lightning/trainer/properties.py [hunk 13]
```diff
@@ -62,13 +68,18 @@ class TrainerProperties(ABC):
     _weights_save_path: str
 
     accelerator_connector: AcceleratorConnector
+    accumulate_grad_batches: int
```
Comment (awaelchli) [inline pytorch_lightning/trainer/properties.py:71]: reviewer: the loops are accessing these attributes, so adding them here with types

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 14]
```diff
@@ -159,6 +161,10 @@ def _run_optimization(
             # if no result, user decided to skip optimization
             # otherwise update running loss + reset accumulated loss
             # TODO: find proper way to handle updating running loss
+            assert self.trainer.fit_loop is not None
+            assert self.trainer.fit_loop.epoch_loop is not None
+            assert self.trainer.fit_loop.epoch_loop.batch_loop is not None
+            assert result.loss is not None
             self.trainer.fit_loop.epoch_loop.batch_loop._update_running_loss(result.loss)
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:168]: reviewer: this looks ridiculous I know. if you have suggestions how to get around this please let me know. Related to that is the TODO above, which is discussed here: #9220

File: pytorch_lightning/core/lightning.py [hunk 15]
```diff
@@ -550,7 +550,7 @@ def __check_allowed(v: Any, name: str, value: Any) -> None:
     def __to_tensor(self, value: numbers.Number) -> torch.Tensor:
         return torch.tensor(value, device=self.device)
 
-    def log_grad_norm(self, grad_norm_dict: Dict[str, torch.Tensor]) -> None:
+    def log_grad_norm(self, grad_norm_dict: Dict[str, float]) -> None:
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:553]: This doesnt seem right, i'll have to check :) reminder to myself.

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 16]
```diff
@@ -206,14 +214,16 @@ def zero_grad_fn():
         if not self._skip_backward and is_first_batch_to_accumulate:
             return zero_grad_fn
 
+        return None
```
Comment (carmocca) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:217]: Can you reestructure this function?

```python

if _skip_backward:
    return


is_first_batch_to_accumulate = batch_idx % self.trainer.accumulate_grad_batches == 0
if is_first_batch_to_accumulate:
    def zero_grad_fn():
        self._on_before_zero_grad(optimizer)
        self._optimizer_zero_grad(batch_idx, optimizer, opt_idx)

    return zero_grad_fn
```

Comment (awaelchli) [reply pytorch_lightning/loops/optimizer/optimizer_loop.py]: done, but need the `return None` explicitly because mypy is strict

File: pytorch_lightning/core/lightning.py (hunk 15 above)
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:553]: the origin of the float values is the [grad_norm](https://github.com/PyTorchLightning/pytorch-lightning/blob/912fd3113143915d3d56f34ec44888911e212926/pytorch_lightning/utilities/grads.py#L51) utility function. This function is called in the loops and then the dict of floats propagates to the LightningModule where I had too change the type

**Review by tchaton — APPROVED** [review]
Looks great !

**Review by SeanNaren — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
